File: mockserver-client-php/src/Llm/StreamingPhysics.php
<?php

declare(strict_types=1);

namespace MockServer\Llm;

use InvalidArgumentException;
use MockServer\Delay;

/**
 * Controls the timing physics of a streamed (SSE) completion
 * (mirrors org.mockserver.model.StreamingPhysics).
 *
 * {@code timeToFirstToken} serialises as a Delay: {@code {timeUnit, value}}.
 */
class StreamingPhysics implements \JsonSerializable
{
    private ?Delay $timeToFirstToken = null;
    private ?int $tokensPerSecond = null;
    private ?float $jitter = null;
    private ?int $seed = null;

    /**
     * Static factory mirroring {@code StreamingPhysics.streamingPhysics()}.
     */
    public static function streamingPhysics(): self
    {
        return new self();
    }

    /**
     * Static factory mirroring {@code StreamingPhysics.tokensPerSecond(n)}.
     */
    public static function tokensPerSecond(int $tokensPerSecond): self
    {
        return (new self())->withTokensPerSecond($tokensPerSecond);
    }

    /**
     * Static factory mirroring {@code StreamingPhysics.jitter(j)}.
     */
    public static function jitter(float $jitter): self
    {
        return (new self())->withJitter($jitter);
    }

    /**
     * Build a Delay representing time-to-first-token.
     */
    public static function timeToFirstToken(int $value, string $timeUnit = 'MILLISECONDS'): Delay
    {
        return new Delay($timeUnit, $value);
    }

    public function withTimeToFirstToken(Delay $delay): self
    {
        $this->timeToFirstToken = $delay;
        return $this;
    }

    public function withTokensPerSecond(int $tokensPerSecond): self
    {
        if ($tokensPerSecond < 1 || $tokensPerSecond > 10000) {
            throw new InvalidArgumentException('tokensPerSecond must be between 1 and 10000');
        }
        $this->tokensPerSecond = $tokensPerSecond;
        return $this;
    }

    public function withJitter(float $jitter): self
    {
        if ($jitter < 0.0 || $jitter > 1.0) {
            throw new InvalidArgumentException('jitter must be between 0.0 and 1.0');
        }
        $this->jitter = $jitter;
        return $this;
    }

    public function withSeed(int $seed): self
    {
        $this->seed = $seed;
        return $this;
    }

    /**
     * @return array<string, mixed>
     */
    public function jsonSerialize(): array
    {
        return $this->toArray();
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $data = [];
        if ($this->timeToFirstToken !== null) {
            $data['timeToFirstToken'] = $this->timeToFirstToken->toArray();
        }
        if ($this->tokensPerSecond !== null) {
            $data['tokensPerSecond'] = $this->tokensPerSecond;
        }
        if ($this->jitter !== null) {
            $data['jitter'] = $this->jitter;
        }
        if ($this->seed !== null) {
            $data['seed'] = $this->seed;
        }
        return $data;
    }
}

File: mockserver/mockserver-core/src/main/java/org/mockserver/model/StreamingPhysics.java
package org.mockserver.model;

import java.util.Objects;

public class StreamingPhysics extends ObjectWithJsonToString {
    private int hashCode;
    private Delay timeToFirstToken;
    private Integer tokensPerSecond;
    private Double jitter;
    private Long seed;
    private Boolean subwordStreaming;

    public static StreamingPhysics streamingPhysics() {
        return new StreamingPhysics();
    }

    public StreamingPhysics withTimeToFirstToken(Delay timeToFirstToken) {
        this.timeToFirstToken = timeToFirstToken;
        this.hashCode = 0;
        return this;
    }

    public Delay getTimeToFirstToken() {
        return timeToFirstToken;
    }

    public StreamingPhysics withTokensPerSecond(Integer tokensPerSecond) {
        if (tokensPerSecond != null && (tokensPerSecond < 1 || tokensPerSecond > 10000)) {
            throw new IllegalArgumentException("tokensPerSecond must be between 1 and 10000");
        }
        this.tokensPerSecond = tokensPerSecond;
        this.hashCode = 0;
        return this;
    }

    public Integer getTokensPerSecond() {
        return tokensPerSecond;
    }

    public StreamingPhysics withJitter(Double jitter) {
        if (jitter != null && (jitter < 0.0 || jitter > 1.0)) {
            throw new IllegalArgumentException("jitter must be between 0.0 and 1.0");
        }
        this.jitter = jitter;
        this.hashCode = 0;
        return this;
    }

    public Double getJitter() {
        return jitter;
    }

    public StreamingPhysics withSeed(Long seed) {
        this.seed = seed;
        this.hashCode = 0;
        return this;
    }

    public Long getSeed() {
        return seed;
    }

    /**
     * Controls streaming delta granularity. Subword-sized deltas (a lightweight
     * BPE approximation, see {@code org.mockserver.llm.TokenCounter}) are now the
     * <strong>default</strong>: leaving this unset ({@code null}) — or setting it
     * {@code true} — emits finer, more realistic per-token deltas. Set it explicitly
     * to {@code false} to opt back into the legacy whole-word (whitespace-boundary)
     * splitting. The per-event timing math is unchanged — each delta is still one
     * physics event — so subword mode simply produces more, smaller events, closer
     * to a real provider's per-token stream (and therefore a slightly longer total
     * stream duration for the same {@code tokensPerSecond}).
     */
    public StreamingPhysics withSubwordStreaming(Boolean subwordStreaming) {
        this.subwordStreaming = subwordStreaming;
        this.hashCode = 0;
        return this;
    }

    public Boolean getSubwordStreaming() {
        return subwordStreaming;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        if (hashCode() != o.hashCode()) {
            return false;
        }
        StreamingPhysics that = (StreamingPhysics) o;
        return Objects.equals(timeToFirstToken, that.timeToFirstToken) &&
            Objects.equals(tokensPerSecond, that.tokensPerSecond) &&
            Objects.equals(jitter, that.jitter) &&
            Objects.equals(seed, that.seed) &&
            Objects.equals(subwordStreaming, that.subwordStreaming);
    }

    @Override
    public int hashCode() {
        if (hashCode == 0) {
            hashCode = Objects.hash(timeToFirstToken, tokensPerSecond, jitter, seed, subwordStreaming);
        }
        return hashCode;
    }
}
